### ISSUE TITLE
launch CPU's VMSA can be in the `MEMORY_MAP`

### ISSUE THREAD

**Issue by Freax13** [coconut-svsm/svsm#22]
We already ensure that no SVSM kernel memory is part of `MEMORY_MAP` provided by the host:
https://github.com/coconut-svsm/svsm/blob/e9b75d58f2ffb530f87b6a51d775deec1d0231bc/src/mm/memory.rs#L23-L28

We should also ensure here (and possible elsewhere) that the launch CPU's VMSA at `0xFFF_FFFF_F000` is not confused with regular, usable memory.

For the record here's the function in the host that executes the launch update for the launch CPU:
https://github.com/coconut-svsm/linux/blob/462abfe0ed20885795ea251320cf63d271bf3216/arch/x86/kvm/svm/sev.c#L2316-L2346

I'm not aware of any way to exploit this, but it seems reasonable to protect the launch CPU's VMSA.

**Comment by joergroedel** [comment]
Hi Tom, thanks for looking into this. I don't get from the code snippet you linked how the launch CPUs VMSA gets mapped into the guests physical address space. I see that it gets assigned the HPA as GPA, but not the mapping itself. Can you elaborate a bit on that?

**Comment by Freax13** [comment]
It may be the case that the VMSA is not actually mapped into the guest physical address space, but the launch update causes the page to be pvalidated, so if a malicious hypervisor wanted to map it into the guest physical address space, it could.

There are also double-pvalidate concerns because the VMSA's GPA isn't tracked as pvalidated in the bitmap.

**Comment by joergroedel** [comment]
Hmm, yes, this is a problem. We need to discuss this with AMD. I think it should be implemented in a way that the launch VMSA is actually mapped into guest physical address space. Otherwise what happens is that the HV has 2 pages valid for the same GPA and can use that to attack the guest. Thanks for bringing this up.

**Comment by joergroedel** [comment]
@tlendacky, can you have a look please? I think the reporter found an issue with how the launch-cpu VMSA is set up. This is not related to SVSM, but to SNP host support.

**Comment by tlendacky** [comment]
I don't think this is any different than a LAUNCH_UPDATE_DATA for non-VMSA pages. The guest needs to know what has been validated already when being launched, if it is part of its memory map, and not try to PVALIDATE that again.

The RMPUPDATE uses -1 for the GPA for VMSA pages, which becomes 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit). So lets look at the following scenarios:

The HV maps the VMSA page at some (random) address in the guest memory
- PVALIDATE will generate a nested page fault because the GPA being validated does not match the GPA in the RMP table
- Forward progress relies on the HV to do an RMPUPDATE to the new GPA, but the VALIDATED bit is maintained
- PVALIDATE will indicate that the page has already been validated, the guest should terminate

The HV maps the VMSA page at 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit)
- If the guest never references this memory (since it will most likely be outside of the e820 map for the guest), no issue.
- Otherwise, the guest would think that it needs to PVALIDATE that memory
- PVALIDATE will indicate that the page has already been validated, the guest should terminate

On the last scenario, I'm not even sure that is possible because of the reduced physical addressing (the c-bit), in which case the first scenario applies.

Is there a different concern I'm missing or does that cover it?

**Comment by Freax13** [comment]
> The RMPUPDATE uses -1 for the GPA for VMSA pages, which becomes 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit). So lets look at the following scenarios:

Are you sure about those addresses? When implementing attestation I had to use `0xFFF_FFFF_F000` (-1 as u32 shifted to the left by 12) to get the correct result. AFAICT this is because it's the GFN that's -1, so that explains the GPA.
  
> The HV maps the VMSA page at 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit)
> 
>   * If the guest never references this memory (since it will most likely be outside of the e820 map for the guest), no issue.

Currently the COCONUT SVSM gets the e820 map from the host, so this is possible.
 
>   * Otherwise, the guest would think that it needs to PVALIDATE that memory
> 
>   * PVALIDATE will indicate that the page has already been validated, the guest should terminate

`PVALIDATE` will not necessarily report that if the host doesn't map the VMSA's GFN to the VMSA's PFN at the time that `PVALIDATE` is executed. If it maps it to some other PFN the `PVALIDATE` will succeed and the same GFN will have two valid RMP entries for two different PFNs (bad!).

**Comment by tlendacky** [comment]
> > The RMPUPDATE uses -1 for the GPA for VMSA pages, which becomes 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit). So lets look at the following scenarios:
> 
> Are you sure about those addresses? When implementing attestation I had to use `0xFFF_FFFF_F000` (-1 as u32 shifted to the left by 12) to get the correct result. AFAICT this is because it's the GFN that's -1, so that explains the GPA.

Attestation measurements are done by the PSP and it determines how to use the input. I updated the previous comment, with what is stored in the RMP.

> 
> > The HV maps the VMSA page at 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit)
> > 
> > * If the guest never references this memory (since it will most likely be outside of the e820 map for the guest), no issue.
> 
> Currently the COCONUT SVSM gets the e820 map from the host, so this is possible.
> 
> > * Otherwise, the guest would think that it needs to PVALIDATE that memory
> > * PVALIDATE will indicate that the page has already been validated, the guest should terminate
> 
> `PVALIDATE` will not necessarily report that if the host doesn't map the VMSA's GFN to the VMSA's PFN at the time that `PVALIDATE` is executed. If it maps it to some other PFN the `PVALIDATE` will succeed and the same GFN will have two valid RMP entries for two different PFNs (bad!).

If the guest is mapping the VMSA address then it must know that it is a VMSA address and was already validated and shouldn't be trying to do a PVALIDATE again.

**Comment by Freax13** [comment]
> > > The RMPUPDATE uses -1 for the GPA for VMSA pages, which becomes 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit). So lets look at the following scenarios:
> > 
> > 
> > Are you sure about those addresses? When implementing attestation I had to use `0xFFF_FFFF_F000` (-1 as u32 shifted to the left by 12) to get the correct result. AFAICT this is because it's the GFN that's -1, so that explains the GPA.
> 
> Attestation measurements are done by the PSP (SEV firmware) and it determines how to use the input. I updated the previous comment with what the RMPUPDATE instruction stores in the RMP.

Okay, so I just checked again, you're right. The RMP update is done with all bits set, not just 32 or 44, and the firmware doesn't use some of the upper bits. To quote the spec (SEV Secure Nested Paging Firmware ABI Specification, 8.17 SNP_LAUNCH_UPDATE):
> Note that the guest physical address space is limited according to CPUID Fn80000008_EAX and
thus the GPAs used by the firmware in measurement calculation are equally limited. 

The one thing I don't understand is why the measured address would be `0xFFF_FFFF_F000`.  The limits mentioned are 48 bits on my system, so I don't understand why the address is truncated at just 44 bits.

> > > The HV maps the VMSA page at 0xFFFF_FFFF_F000 (for 48-bit) or 0xF_FFFF_FFFF_F000 (for 52-bit)
> > > 
> > > * If the guest never references this memory (since it will most likely be outside of the e820 map for the guest), no issue.
> > 
> > 
> > Currently the COCONUT SVSM gets the e820 map from the host, so this is possible.
> > > * Otherwise, the guest would think that it needs to PVALIDATE that memory
> > > * PVALIDATE will indicate that the page has already been validated, the guest should terminate
> > 
> > 
> > `PVALIDATE` will not necessarily report that if the host doesn't map the VMSA's GFN to the VMSA's PFN at the time that `PVALIDATE` is executed. If it maps it to some other PFN the `PVALIDATE` will succeed and the same GFN will have two valid RMP entries for two different PFNs (bad!).
> 
> If the guest is mapping the VMSA address then it must know that it is a VMSA address and was already validated and shouldn't be trying to do a PVALIDATE again.

So we are in agreement that the SVSM should be aware of the VMSA gPA and should actively avoid pvalidating it or mapping it as usable memory?

**Comment by tlendacky** [comment]
> Are you sure about those addresses? When implementing attestation I had to use `0xFFF_FFFF_F000` (-1 as u32 shifted to the left by 12) to get the correct result. AFAICT this is because it's the GFN that's -1, so that explains the GPA.

The SEV firmware will use the GPA value that is stored in the RMP, so on Milan I would expect the value you would use is 0xFFFF_FFFF_F000 and not 0xFFF_FFFF_F000 and on Genoa I would expect 0x7_FFFF_FFFF_F000.

The RMP GPA is stored as bits GPA[50:12] & <System PA Mask> (48-bit on Milan, so mask is 0xFFFF_FFFF_FFFF and 52-bit on Genao, so mask is 0xF_FFFF_FFFF_FFFF), effectively a GFN. So the GPA used by SEV firmware is the 39-bit RMP value left shifted 12-bits.

I dumped the RMP entry for a VMSA page on both Milan and Genoa.

Milan:
  SEV-SNP: RMPEntry paddr 0x145ba4000 [assigned=1 immutable=0 pagesize=0 gpa=0xfffffffff asid=1 vmsa=1 validated=1]

Genoa:
  SEV-SNP: RMPEntry paddr 0x1c9f36000 [assigned=1 immutable=0 pagesize=0 gpa=0x7fffffffff asid=1 vmsa=1 validated=1]

**Comment by Freax13** [comment]
Nevermind, I missed one digit. You're correct it's using all the bits and behaving exactly like specified. My bad.

**Comment by tlendacky** [comment]
> So we are in agreement that the SVSM should be aware of the VMSA gPA and should actively avoid pvalidating it or mapping it as usable memory?

Yes, and the attestation report would indicate if the HV tried to use a different/wrong value for the VMSA gPA.
